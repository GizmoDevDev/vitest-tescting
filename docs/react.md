# Тестирование React

Тестирование React будем проводить с помощью `React Testing Library`

## Настройка
1. `yarn add -D jsdom` устанавливаем jsdom для доступа к html в vitest
2. Добавляем в настройку vite или vitest jsdom в качестве окружения
```js
test: {
    environment: 'jsdom',
},
```
3. `yarn add -D @testing-library/react @testing-library/jest-dom` подключаем саму библиотеку `React Testing Library`
4. `yarn add -D @testing-library/user-event` добавляем библиотеку, чтобы имитировать действия пользователей
5. Создаем файл setup, чтобы не писать настройки для каждого файла отдельно со следующим содержимым:
```js
//setup.js
import { expect, afterEach } from 'vitest';
import { cleanup } from '@testing-library/react';
import matchers from '@testing-library/jest-dom/matchers';

// extends Vitest's expect method with methods from react-testing-library
expect.extend(matchers);

// runs a cleanup after each test case (e.g. clearing jsdom)
afterEach(() => {
  cleanup();
});
```
6. Дорабатываем настройки vitest:
```js
  test: {
    globals: true,
    environment: 'jsdom',
    setupFiles: 'setup.js', // Нужно указать весь путь для файла setup.js
  },
```