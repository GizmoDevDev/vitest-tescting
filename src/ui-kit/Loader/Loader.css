.loader_size-s {
  width: 20px;
  height: 20px;
  border: 3px solid #000000;
  border-left: 3px solid #ffffff;
}
.loader_size-m {
  width: 40px;
  height: 40px;
  border: 4.5px solid #000000;
  border-left: 4.5px solid #ffffff;
}
.loader_size-l {
  width: 60px;
  height: 60px;
  border: 6px solid #000000;
  border-left: 6px solid #ffffff;
}

.loader {
  display: inline-block;
  vertical-align: middle;
  position: relative;
  border-radius: 50%;
  box-sizing: border-box;
  animation: 2s linear 0s normal none infinite running loader;
  -webkit-animation: 2s linear 0s normal none infinite running loader;
}

@keyframes loader {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
@-webkit-keyframes loader {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
