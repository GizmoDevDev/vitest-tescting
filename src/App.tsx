import {NumberInput} from "./ui-kit/NumberInput/NumberInput";
import './App.css';
import {useEffect, useState} from "react";
import {Loader} from "./ui-kit/Loader/Loader";
function App() {
  const [rub, setRub] = useState(0)
  const [usd, setUsd] = useState(0)
  const [exchangeRate, setExchangeRate] = useState(0)
  const [isLoading, setIsLoading] = useState(true)

  useEffect(() => {
    fetch("http://www.randomnumberapi.com/api/v1.0/random?min=70&max=100&count=1")
      .then((resp) => resp.json())
      .then((data) => {
        setExchangeRate(data[0])
      })
      .catch((e) => console.error(e))
      .finally(() => setIsLoading(false))
  }, [])

  if (isLoading) return <Loader />

  const handleChangeRub = (v: number) => {
    setRub(v);
    setUsd(v / exchangeRate);
  }

  const handleChangeUsd = (v: number) => {
    setUsd(v);
    setRub(v * exchangeRate);
  }

  return (
    <div>
      <NumberInput value={rub} onChange={handleChangeRub} />
      <div>Exchange rate: {exchangeRate}</div>
      <NumberInput value={usd} onChange={handleChangeUsd} />
    </div>
  )
}

export default App
