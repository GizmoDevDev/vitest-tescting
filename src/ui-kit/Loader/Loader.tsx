import clsx from 'clsx';
import React from 'react';
import './Loader.css';

export enum LoaderSize {
  s = 's',
  m = 'm',
  l = 'l',
}

export type LoaderProps = {
  loading?: boolean;
  size?: LoaderSize | keyof typeof LoaderSize;
  className?: string;
};

export const Loader = ({
  loading = true,
  size = LoaderSize.l,
  className,
}: LoaderProps) => {
  if (!loading) return null;
  return <div className={clsx('loader', `loader_size-${size}`, className)} />
};
