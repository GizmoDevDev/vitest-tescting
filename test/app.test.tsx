import {afterEach, beforeEach, describe, expect, Mock, test, vi} from "vitest";
import {render, screen} from '@testing-library/react';
import App from "../src/App";
import { mockLoaderTestId } from './mock/Loader';
import {setImmediate} from 'timers'
import userEvent from "@testing-library/user-event";

vi.mock('../src/ui-kit/Loader/Loader',  () => ({
  Loader: () => <div data-testId={mockLoaderTestId} />
}) )

describe('тестирование экрана с конвертацией валют', () => {
  beforeEach(() => {
    global.fetch = vi.fn();
  })
  afterEach(() => {
    vi.clearAllMocks()
  })

  test('Loading before fetch execute', () => {
    (fetch as Mock).mockResolvedValue({json: () => Promise.resolve([54])})
    render(<App />)
    const loader = screen.getByTestId(mockLoaderTestId);
    expect(loader).not.toBe(null)
  })

  test('Loading doesnt exist after fetch execute', async () => {
    (fetch as Mock).mockResolvedValue({json: () => Promise.resolve([54])})
    render(<App />)
    await new Promise(setImmediate)
    expect(() => screen.getByTestId(mockLoaderTestId)).toThrow();
  })

  test('Change rub input', async () => {
    (fetch as Mock).mockResolvedValue({json: () => Promise.resolve([54])})
    const { container, rerender } = render(<App />)
    await new Promise(setImmediate)
    rerender(<App />)
    const input = container.querySelector('input');
    if (!input) {
      test.fails('Doesnt exist input')
      return
    }
    await userEvent.type(input, '54')
    expect(container).toMatchSnapshot()
  })

  test('Change usd input', async () => {
    (fetch as Mock).mockResolvedValue({json: () => Promise.resolve([54])})
    const { container, rerender } = render(<App />)
    await new Promise(setImmediate)
    rerender(<App />)
    const input = container.querySelectorAll('input')[1];
    if (!input) {
      test.fails('Doesnt exist input')
      return
    }
    await userEvent.type(input, '54')
    expect(container).toMatchSnapshot()
  })
})