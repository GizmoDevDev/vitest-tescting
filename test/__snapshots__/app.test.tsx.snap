// Vitest Snapshot v1

exports[`тестирование экрана с конвертацией валют > Change rub input 1`] = `
<div>
  <div>
    <input
      type="number"
      value="0"
    />
    <div>
      Exchange rate: 
      54
    </div>
    <input
      type="number"
      value="1"
    />
  </div>
</div>
`;

exports[`тестирование экрана с конвертацией валют > Change usd input 1`] = `
<div>
  <div>
    <input
      type="number"
      value="2916"
    />
    <div>
      Exchange rate: 
      54
    </div>
    <input
      type="number"
      value="0"
    />
  </div>
</div>
`;
